File: features/cross_platform/lib/features/core/domain/entities/product.dart
import 'package:frosy_pine/features/core/domain/entities/brand.dart';
import 'package:frosy_pine/features/core/domain/entities/category.dart';

final class Product {
  Product({
    required this.id,
    required this.name,
    required this.brand,
    required this.category,
  });

  final String id;
  final String name;
  final Brand brand;
  final Category category;
}

final class LazyProduct {
  LazyProduct({
    required this.id,
    required this.name,
    required this.brandId,
    required this.categoryId,
  });

  final String id;
  final String name;
  final String brandId;
  final String categoryId;

  Product toProduct(Brand brand, Category category) =>
      Product(id: id, name: name, brand: brand, category: category);
}

File: expense_tracking/src/domain/entities/product.rs
use uuid::Uuid;
use uuid_b64::UuidB64;

use crate::domain::entities::Brand;
use crate::domain::entities::Category;

#[derive(Default, Debug, Clone, PartialEq)]
pub struct Product {
    pub id: UuidB64,
    pub name: String,
    pub brand: Brand,
    pub category: Category,
}

impl Product {
    pub fn new(id: Option<UuidB64>, name: String, brand: Brand, category: Category) -> Self {
        Self {
            id: id.unwrap_or_else(|| UuidB64::from(Uuid::new_v4())),
            name,
            brand,
            category,
        }
    }
}
